### PR TITLE
add test for trec covid scripts

### PR DESCRIPTION
**Pull request by yuki617**
add test for testing the two downloaed qrels files;
add test for the script to generate runs file and evaluate scores

### CODE REVIEW

File: integrations/test_trec_covid_r3.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+class TestSearchIntegration(unittest.TestCase):
+    def setUp(self):
+        self.round3_runs = {
+            'https://raw.githubusercontent.com/castorini/anserini/master/src/main/resources/topics-and-qrels/qrels.covid-round3-cumulative.txt':
+                'dfccc32efd58a8284ae411e5c6b27ce9',
+            'https://raw.githubusercontent.com/castorini/anserini/master/src/main/resources/topics-and-qrels/qrels.covid-round4-cumulative.txt':
+                '7a5c27e8e052c49ff72d557051825973'
+        }
+        self.tmp = f'../trec-covid-r3/check_data'
```
Comment (lintool) [inline integrations/test_trec_covid_r3.py:33]: put in `runs/`? Add a random number to directory to reduce change of collisions.

File: integrations/test_trec_covid_r3.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            print('')
+
+
+    def test_round3_score_runs(self):
+        os.system(f'python3 ../trec-covid-r3/ranker.py \
+            -alpha 0.6 \
+            -clf lr \
+            -vectorizer tfidf \
+            -trec_covid_home ../trec-covid-r3 \
+            -base ../trec-covid-r3/data/covidex.r4.d2q.duot5 \
```
Comment (lintool) [inline integrations/test_trec_covid_r3.py:56]: Use the prebuilt indexes!

Comment (lintool) [reply integrations/test_trec_covid_r3.py:56]: Also, topic and qrels should be in ` tools/topics-and-qrels/`

File: integrations/test_trec_covid_r3.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            -qrels {self.tmp}/qrels.covid-round3-cumulative.txt \
+            -index ../trec-covid-r3/data/lucene-index-cord19-abstract-2020-06-19 \
+            -tag ../trec-covid-r3/data/covidex.r4.d2q.duot5.lr')
```
Comment (lintool) [inline integrations/test_trec_covid_r3.py:59]: you need an actual check of MAP, NDCG, or whatever...

File: integrations/test_trec_covid_r3.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+
+    # def tearDown(self):
```
Comment (lintool) [inline integrations/test_trec_covid_r3.py:62]: we _do_ want cleanup.

**Review by lintool — CHANGES_REQUESTED** [review]
Various comments...

File: integrations/test_trec_covid_r3.py (hunk 4 above)
Comment (yuki617) [reply integrations/test_trec_covid_r3.py]: All changes are done. Thanks!

**Comment by lintool** [comment]
Doesn't work for me:

```
$ python -m unittest integrations/test_trec_covid_r3.py 
E
======================================================================
ERROR: test_bm25 (integrations.test_trec_covid_r3.TestSearchIntegration)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jimmylin/workspace/pyserini/integrations/test_trec_covid_r3.py", line 47, in setUp
    for url in self.round3_runs:
AttributeError: 'TestSearchIntegration' object has no attribute 'round3_runs'

----------------------------------------------------------------------
Ran 1 test in 0.001s

FAILED (errors=1)
```

**Comment by yuki617** [comment]
> Doesn't work for me:
> 
> ```
> $ python -m unittest integrations/test_trec_covid_r3.py 
> E
> ======================================================================
> ERROR: test_bm25 (integrations.test_trec_covid_r3.TestSearchIntegration)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/Users/jimmylin/workspace/pyserini/integrations/test_trec_covid_r3.py", line 47, in setUp
>     for url in self.round3_runs:
> AttributeError: 'TestSearchIntegration' object has no attribute 'round3_runs'
> 
> ----------------------------------------------------------------------
> Ran 1 test in 0.001s
> 
> FAILED (errors=1)
> ```

Should works now. It's because a typo in function name.

File: integrations/test_trec_covid_r3.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+            download_url(url, self.tmp, md5=self.round4_runs[url], force=True)
+            self.assertTrue(os.path.exists(os.path.join(self.tmp, filename)))
+            print('')
+
+
+    def test_bm25(self):
+
+        prebuilt_index_path = download_prebuilt_index('trec-covid-r4-abstract')
+        os.system(f'python ../trec-covid-r3/ranker.py \
```
Comment (lintool) [inline integrations/test_trec_covid_r3.py:60]: This file doesn't exist, so I can't run the integration test...
Should probably be in `scripts/`

**Review by lintool — CHANGES_REQUESTED** [review]
Still doesn't work.

```
python: can't open file '../trec-covid-r3/ranker.py': [Errno 2] No such file or directory
```

File: integrations/test_trec_covid_r3.py (hunk 5 above)
Comment (yuki617) [reply integrations/test_trec_covid_r3.py:60]: The ranker.py should be under trec-covid-r3. Can I know where did you put the trec-covid-r3?

Comment (yuki617) [reply integrations/test_trec_covid_r3.py:60]: Also, to run the script, we should keep a ../trec-covid-r3/data/covidex.r4.d2q.duot5 and ../trec-covid-r3/data/covidex.r4.d2q.duot5.lr files. Should I assume the user contain these two files or download them in the setup?

Comment (lintool) [reply integrations/test_trec_covid_r3.py:60]: You can download the runs to `runs/`

Comment (yuki617) [reply integrations/test_trec_covid_r3.py]: Done it. Also pls pull the newest trec-covid-r3 for testing. Thanks a lot!

**Review by lintool — CHANGES_REQUESTED** [review]
Still doesn't work... you haven't checked in missing files...

```
$ python -m unittest integrations/test_trec_covid_r3.py
Downloading https://ir.nist.gov/covidSubmit/archive/round4/covidex.r4.d2q.duot5.gz to runs/covidex.r4.d2q.duot5.gz...
covidex.r4.d2q.duot5.gz: 768kB [00:00, 1.11MB/s]                                                                                                                                                                                                                                    
Verifying stored run at https://raw.githubusercontent.com/castorini/anserini/master/src/main/resources/topics-and-qrels/qrels.covid-round3-cumulative.txt...
Downloading https://raw.githubusercontent.com/castorini/anserini/master/src/main/resources/topics-and-qrels/qrels.covid-round3-cumulative.txt to tmp2184/qrels.covid-round3-cumulative.txt...
qrels.covid-round3-cumulative.txt: 536kB [00:00, 1.56MB/s]                                                                                                                                                                                                                          

Verifying stored run at https://raw.githubusercontent.com/castorini/anserini/master/src/main/resources/topics-and-qrels/qrels.covid-round4-cumulative.txt...
Downloading https://raw.githubusercontent.com/castorini/anserini/master/src/main/resources/topics-and-qrels/qrels.covid-round4-cumulative.txt to tmp2184/qrels.covid-round4-cumulative.txt...
qrels.covid-round4-cumulative.txt: 752kB [00:00, 2.49MB/s]                                                                                                                                                                                                                          

/Users/jimmylin/.cache/pyserini/indexes/lucene-index-cord19-abstract-2020-06-19.029bd55daba8800fbae2be9e5fcd7b33 already exists, skipping download.
python3: can't open file '../../trec-covid-r3/ranker.py': [Errno 2] No such file or directory
EE
======================================================================
ERROR: test_bm25 (integrations.test_trec_covid_r3.TestSearchIntegration)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jimmylin/workspace/pyserini/integrations/test_trec_covid_r3.py", line 77, in test_bm25
    with open(f'{self.tmp}/output.json') as json_file:
FileNotFoundError: [Errno 2] No such file or directory: 'tmp2184/output.json'

======================================================================
ERROR: test_bm25 (integrations.test_trec_covid_r3.TestSearchIntegration)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jimmylin/workspace/pyserini/integrations/test_trec_covid_r3.py", line 87, in tearDown
    os.remove('runs/covidex.r4.d2q.duot5.lr.tfidf.R12.A0.6.txt')
FileNotFoundError: [Errno 2] No such file or directory: 'runs/covidex.r4.d2q.duot5.lr.tfidf.R12.A0.6.txt'

----------------------------------------------------------------------
Ran 1 test in 1.426s

FAILED (errors=2)
```

**Comment by yuki617** [comment]
> Still doesn't work... you haven't checked in missing files...
> 
> ```
> $ python -m unittest integrations/test_trec_covid_r3.py
> Downloading https://ir.nist.gov/covidSubmit/archive/round4/covidex.r4.d2q.duot5.gz to runs/covidex.r4.d2q.duot5.gz...
> covidex.r4.d2q.duot5.gz: 768kB [00:00, 1.11MB/s]                                                                                                                                                                                                                                    
> Verifying stored run at https://raw.githubusercontent.com/castorini/anserini/master/src/main/resources/topics-and-qrels/qrels.covid-round3-cumulative.txt...
> Downloading https://raw.githubusercontent.com/castorini/anserini/master/src/main/resources/topics-and-qrels/qrels.covid-round3-cumulative.txt to tmp2184/qrels.covid-round3-cumulative.txt...
> qrels.covid-round3-cumulative.txt: 536kB [00:00, 1.56MB/s]                                                                                                                                                                                                                          
> 
> Verifying stored run at https://raw.githubusercontent.com/castorini/anserini/master/src/main/resources/topics-and-qrels/qrels.covid-round4-cumulative.txt...
> Downloading https://raw.githubusercontent.com/castorini/anserini/master/src/main/resources/topics-and-qrels/qrels.covid-round4-cumulative.txt to tmp2184/qrels.covid-round4-cumulative.txt...
> qrels.covid-round4-cumulative.txt: 752kB [00:00, 2.49MB/s]                                                                                                                                                                                                                          
> 
> /Users/jimmylin/.cache/pyserini/indexes/lucene-index-cord19-abstract-2020-06-19.029bd55daba8800fbae2be9e5fcd7b33 already exists, skipping download.
> python3: can't open file '../../trec-covid-r3/ranker.py': [Errno 2] No such file or directory
> EE
> ======================================================================
> ERROR: test_bm25 (integrations.test_trec_covid_r3.TestSearchIntegration)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/Users/jimmylin/workspace/pyserini/integrations/test_trec_covid_r3.py", line 77, in test_bm25
>     with open(f'{self.tmp}/output.json') as json_file:
> FileNotFoundError: [Errno 2] No such file or directory: 'tmp2184/output.json'
> 
> ======================================================================
> ERROR: test_bm25 (integrations.test_trec_covid_r3.TestSearchIntegration)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/Users/jimmylin/workspace/pyserini/integrations/test_trec_covid_r3.py", line 87, in tearDown
>     os.remove('runs/covidex.r4.d2q.duot5.lr.tfidf.R12.A0.6.txt')
> FileNotFoundError: [Errno 2] No such file or directory: 'runs/covidex.r4.d2q.duot5.lr.tfidf.R12.A0.6.txt'
> 
> ----------------------------------------------------------------------
> Ran 1 test in 1.426s
> 
> FAILED (errors=2)
> ```

The previous script must be called inside integration, otherwise there will be missing files. I fixed the problem, it can be called both in pyserini root and integration folder now. Please try to run it again. Thanks for your patience and time. Also, please update the trec-covid-r3, the script there is changed too.

**Comment by lintool** [comment]
I still have issues...

```
$ python -m unittest integrations/test_trec_covid_r3.py
['', '/anaconda3/envs/python36/lib/python36.zip', '/anaconda3/envs/python36/lib/python3.6', '/anaconda3/envs/python36/lib/python3.6/lib-dynload', '/Users/jimmylin/.local/lib/python3.6/site-packages', '/anaconda3/envs/python36/lib/python3.6/site-packages', '..']
Downloading https://ir.nist.gov/covidSubmit/archive/round4/covidex.r4.d2q.duot5.gz to ./integrations/runs/covidex.r4.d2q.duot5.gz...
covidex.r4.d2q.duot5.gz: 0.00B [00:00, ?B/s]
E
======================================================================
ERROR: test_bm25 (integrations.test_trec_covid_r3.TestSearchIntegration)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jimmylin/workspace/pyserini/integrations/test_trec_covid_r3.py", line 48, in setUp
    f'{self.pyserini_root}/integrations/runs')
  File "/Users/jimmylin/workspace/pyserini/pyserini/util.py", line 80, in download_url
    urlretrieve(url, filename=destination_path, reporthook=t.update_to)
  File "/anaconda3/envs/python36/lib/python3.6/urllib/request.py", line 258, in urlretrieve
    tfp = open(filename, 'wb')
FileNotFoundError: [Errno 2] No such file or directory: './integrations/runs/covidex.r4.d2q.duot5.gz'

----------------------------------------------------------------------
Ran 1 test in 0.294s

FAILED (errors=1)
```

Relative path is incorrect... but even if I do:

```
$ mkdir integrations/runs
```

Still other path issues:

```
$ python -m unittest integrations/test_trec_covid_r3.py
['', '/anaconda3/envs/python36/lib/python36.zip', '/anaconda3/envs/python36/lib/python3.6', '/anaconda3/envs/python36/lib/python3.6/lib-dynload', '/Users/jimmylin/.local/lib/python3.6/site-packages', '/anaconda3/envs/python36/lib/python3.6/site-packages', '..']
Downloading https://ir.nist.gov/covidSubmit/archive/round4/covidex.r4.d2q.duot5.gz to ./integrations/runs/covidex.r4.d2q.duot5.gz...
covidex.r4.d2q.duot5.gz: 768kB [00:00, 1.43MB/s]                                                                                                                                                                                                                                    
Verifying stored run at https://raw.githubusercontent.com/castorini/anserini/master/src/main/resources/topics-and-qrels/qrels.covid-round3-cumulative.txt...
Downloading https://raw.githubusercontent.com/castorini/anserini/master/src/main/resources/topics-and-qrels/qrels.covid-round3-cumulative.txt to ./integrations/tmp6854/qrels.covid-round3-cumulative.txt...
qrels.covid-round3-cumulative.txt: 536kB [00:00, 1.39MB/s]                                                                                                                                                                                                                          

Verifying stored run at https://raw.githubusercontent.com/castorini/anserini/master/src/main/resources/topics-and-qrels/qrels.covid-round4-cumulative.txt...
Downloading https://raw.githubusercontent.com/castorini/anserini/master/src/main/resources/topics-and-qrels/qrels.covid-round4-cumulative.txt to ./integrations/tmp6854/qrels.covid-round4-cumulative.txt...
qrels.covid-round4-cumulative.txt: 752kB [00:00, 2.49MB/s]                                                                                                                                                                                                                          

/Users/jimmylin/.cache/pyserini/indexes/lucene-index-cord19-abstract-2020-06-19.029bd55daba8800fbae2be9e5fcd7b33 already exists, skipping download.
python: can't open file './../trec-covid-r3/ranker.py': [Errno 2] No such file or directory
EE
======================================================================
ERROR: test_bm25 (integrations.test_trec_covid_r3.TestSearchIntegration)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jimmylin/workspace/pyserini/integrations/test_trec_covid_r3.py", line 81, in test_bm25
    with open(f'{self.tmp}/output.json') as json_file:
FileNotFoundError: [Errno 2] No such file or directory: './integrations/tmp6854/output.json'

======================================================================
ERROR: test_bm25 (integrations.test_trec_covid_r3.TestSearchIntegration)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jimmylin/workspace/pyserini/integrations/test_trec_covid_r3.py", line 90, in tearDown
    os.remove(f'{self.pyserini_root}/integrations/runs/covidex.r4.d2q.duot5.lr.tfidf.R12.A0.6.txt')
FileNotFoundError: [Errno 2] No such file or directory: './integrations/runs/covidex.r4.d2q.duot5.lr.tfidf.R12.A0.6.txt'

----------------------------------------------------------------------
Ran 1 test in 1.310s

FAILED (errors=2)
```

**Comment by lintool** [comment]
Might also want to take a look at this:
https://github.com/castorini/pyserini/blob/master/integrations/test_simplesearcher_check_core17.py#L26

But if I run in IDE via right click, run test, I still get:

```
Testing started at 9:45 PM ...
/anaconda3/envs/python36/bin/python "/Applications/PyCharm CE.app/Contents/plugins/python-ce/helpers/pycharm/_jb_unittest_runner.py" --path /Users/jimmylin/workspace/pyserini/integrations/test_trec_covid_r3.py
Launching unittests with arguments python -m unittest /Users/jimmylin/workspace/pyserini/integrations/test_trec_covid_r3.py in /Users/jimmylin/workspace/pyserini/integrations

['/Users/jimmylin/workspace/pyserini/integrations', '/Users/jimmylin/workspace/pyserini', '/Applications/PyCharm CE.app/Contents/plugins/python-ce/helpers/pycharm', '/anaconda3/envs/python36/lib/python36.zip', '/anaconda3/envs/python36/lib/python3.6', '/anaconda3/envs/python36/lib/python3.6/lib-dynload', '/Users/jimmylin/.local/lib/python3.6/site-packages', '/anaconda3/envs/python36/lib/python3.6/site-packages', '..']
Downloading https://ir.nist.gov/covidSubmit/archive/round4/covidex.r4.d2q.duot5.gz to ../integrations/runs/covidex.r4.d2q.duot5.gz...
covidex.r4.d2q.duot5.gz: 768kB [00:00, 1.40MB/s]                          
Verifying stored run at https://raw.githubusercontent.com/castorini/anserini/master/src/main/resources/topics-and-qrels/qrels.covid-round3-cumulative.txt...
Downloading https://raw.githubusercontent.com/castorini/anserini/master/src/main/resources/topics-and-qrels/qrels.covid-round3-cumulative.txt to ../integrations/tmp9167/qrels.covid-round3-cumulative.txt...
qrels.covid-round3-cumulative.txt: 536kB [00:00, 2.50MB/s]                            

Verifying stored run at https://raw.githubusercontent.com/castorini/anserini/master/src/main/resources/topics-and-qrels/qrels.covid-round4-cumulative.txt...
Downloading https://raw.githubusercontent.com/castorini/anserini/master/src/main/resources/topics-and-qrels/qrels.covid-round4-cumulative.txt to ../integrations/tmp9167/qrels.covid-round4-cumulative.txt...
qrels.covid-round4-cumulative.txt: 752kB [00:00, 4.19MB/s]                           

/Users/jimmylin/.cache/pyserini/indexes/lucene-index-cord19-abstract-2020-06-19.029bd55daba8800fbae2be9e5fcd7b33 already exists, skipping download.
python: can't open file '../../trec-covid-r3/ranker.py': [Errno 2] No such file or directory


Ran 1 test in 1.023s

FAILED (errors=2)
```

**Comment by yuki617** [comment]
I put the required trec-covid-r3 file to folder script. And add script to mkdir a runs folder inside tmp. It should work this time. Thanks!

**Comment by yuki617** [comment]
The scripts for round 3 and round 4 are both changed. Both of them should closely match the score on leaderboard. Thanks!

**Comment by lintool** [comment]
Hi @yuki617 I'm closing this PR. We basically need to rework COVID into what we have for #513 - so probably easier to start over...?

**Pull request closed without merging**